docs: case study on flint cloud (#339)
moja-global/community-website: website/case-studies/flint-cloud.md

---
id:  flint-cloud
title:  FLINT Cloud
slug:  /flint-cloud
description:  Origin and importance of FLINT Cloud
keywords:
- FLINT Modules
- FLINT
- FLINT cloud
- FLINT UI
- Cloud\DevOps
---

> FLINT Cloud is a project being developed to demonstrate the abilities of the [FLINT](https://docs.moja.global/en/master/FLINT/index.html) and its associated services on remote cloud resources. The project aims to build a continuous deployment pipeline to offer FLINT as a SaaS over cloud

Reviewer: ```suggestion
FLINT Cloud is a project being developed to demonstrate the abilities of the [FLINT](https://docs.moja.global/en/master/FLINT/index.html) and its associated services on remote cloud resources. The project aims to build a continuous deployment pipeline that can offer FLINT as a SaaS over cloud.
```

> It provides an abstraction of the underlying implementation while offering the users a simple way to understand the functionality of the FLINT

Reviewer: ```suggestion
It provides an abstraction of the underlying implementation while offering the users a simple way to understand the functionality of the FLINT.
```

## Why is it needed?

> - Today, remote cloud resources are extremely accessible. Providers like Amazon, Google, and Microsoft offer great flexibility and scalability. They also offer very generous free tiers, which will allow users to perform smaller runs.

Reviewer: ```suggestion
- Today, remote cloud resources are extremely accessible. Providers like Amazon, Google, and Microsoft offer great flexibility and scalability. They also offer very generous free tiers that allow users to perform smaller runs.
```

> - Thus, in deploying the FLINT on remote cloud resources the main objective is to increase ease of use and reduce the dependency on local systems, configurations, etc 

Reviewer: ```suggestion
- Thus, in deploying the FLINT on remote cloud resources, the main objective is to increase ease of use and reduce the dependency on local systems, configurations, etc.
```

> - Developers experienced with the FLINTcloud framework will be able to support existing users with CI/CD pipelines, simplify model development and deployment, and leverage auto-scaling for large FLINT runs.

Reviewer: ```suggestion
- Developers who are already experienced with the FLINT cloud framework will be able to support existing users with building CI/CD pipelines, simplifying model development and deployment, and leveraging auto-scaling for large FLINT runs.
```

> - Moja global does not offer FLINT as a Saas on a commercial level, however, through the FLINT Cloud project, they offer deployment templates and workflow training. In January 2022, the organization received a generous credit grant from Microsoft to develop templates for users that wish to deploy tools on [Azure cloud](https://moja.global/moja-global-wins-azure-cloud-grant/)

Reviewer: ```suggestion
- Moja global does not offer FLINT as a Saas on a commercial level, however, through the FLINT Cloud project, they offer deployment templates and workflow training. In January 2022, the organization received a generous credit grant from Microsoft to develop templates for users that wish to deploy tools on [Azure cloud](https://moja.global/moja-global-wins-azure-cloud-grant/).
```

## What is supported currently?

> Currently, the users can run the FLINT and GCBM examples on local and production environments

Reviewer: ```suggestion
Currently users can run the FLINT and GCBM examples on local and production environments.
```

- **Local environment** 

>   - The FLINT and GCBM example docker images can be pulled and built from the Github Container Registry. 

Reviewer: Is it possible to add a link to the GitHub container registry?

>   - The steps to [locally](https://github.com/moja-global/FLINT.Cloud/tree/master/kubernetes/local) deploy the FLINT and GCBM examples on Kubernetes are available 

Reviewer: ```suggestion
  - There are instructions on how to [locally deploy](https://github.com/moja-global/FLINT.Cloud/tree/master/kubernetes/local) the FLINT and GCBM examples on Kubernetes.
```

 
- **Production environment** : 

>  - Using Terraform, an IaaS, deployments can be made on Azure and GCP.

Reviewer: ```suggestion
 - Using Terraform, an IaaS, deployments can be made on Azure and Google Cloud Platform.
```

>  - Steps to deploy FLINT and GCBM examples on the [Azure Kubernetes Service](https://github.com/moja-global/FLINT.Cloud/tree/master/kubernetes/azure) are available

Reviewer: ```suggestion
 - Threre are instructions on how to deploy FLINT and GCBM examples on the [Azure Kubernetes Service](https://github.com/moja-global/FLINT.Cloud/tree/master/kubernetes/azure).
```

### FLINT Example 

> The [FLINT Example](https://github.com/moja-global/FLINT.Cloud/tree/master/local/rest_api_flint.example) demonstrates the Point and the Rothc modules. 

Reviewer: ```suggestion
The [FLINT Example](https://github.com/moja-global/FLINT.Cloud/tree/master/local/rest_api_flint.example) demonstrates the **Point** and **RothC** modules. 
```

> The **Point module** is a simple test module, incorporated to present the concepts of the FLINT such as [Pools, Fluxes, Events and Conservation of mass](https://docs.moja.global/en/master/FLINT/UnderstandingFLINT/MojaFlint/index.html)

Reviewer: ```suggestion
The **Point module** is a simple test module that is used to demonstrate the main concepts of the FLINT such as [Pools, Fluxes, Events and Conservation of mass](https://docs.moja.global/en/master/FLINT/UnderstandingFLINT/MojaFlint/index.html)
```

> More about the [**RothC module**](https://community.moja.global/case-studies/introduction-to-flint-modules#rothc-soil-carbon-module) can be found in the case study on flint modules

Reviewer: ```suggestion
More information about the [**RothC module**](https://community.moja.global/case-studies/introduction-to-flint-modules#rothc-soil-carbon-module) can be found in the case study on FLINT modules.
```

### GCBM Example

> The [GCBM example](https://github.com/moja-global/FLINT.Cloud/tree/master/local/rest_api_gcbm) enables users to understand the entire GCBM simulation lifecycle. A GCBM simulation can be created in a few minutes, with no pre-requisites, apart from having about 5.95 GB for the Docker image. There are endpoints to create a new simulation, upload inputs, start the simulation, and obtain the outputs.

Reviewer: ```suggestion
The [GCBM example](https://github.com/moja-global/FLINT.Cloud/tree/master/local/rest_api_gcbm) enables users to understand the entire GCBM simulation lifecycle. A GCBM simulation can be created in a few minutes and only needs approximately 5.95 GB memory for the Docker image. There are endpoints to create new simulations, upload inputs, start simulations, and obtain the results.
```

> Earlier, the inputs required to run the GCBM simulation were: configuration JSON files, input database and a folder containing the disturbances files (optional), classifier TIFF and JSON files

Reviewer: ```suggestion
Previously, the inputs required to run the GCBM simulation were: 
- Configuration JSON files
- An input database
- Classifier TIFF and JSON files
- A folder containing disturbances files (Optional)
```

> Currently, only the disturbance and classifier TIFF files, along with input database are needed. A default configuration is used to run the simulation

Reviewer: ```suggestion
Currently, only the following is needed:
- Disturbance files
- Classifier TIFF files
- An input database

A default configuration is used to run a simulation.
```

> A future scope is to enable the users to edit configuration files and choose the modules that will be required

Reviewer: ```suggestion
In future, we plan to provide support so that users can edit configuration files and choose which modules will be required for the simulation.
```

## Architecture

> The FLINT is a CLI-based tool. In order to move it into a cloud-based environment, it is packaged into a web server making the use of RESTful API.  

Reviewer: ```suggestion
The FLINT is a CLI-based tool. In order to move it into a cloud-based environment, it is packaged into a web server by using RESTful API.  
```

> A wrapper REST API is created to provide the different endpoints which point to the different operations of the CLI tool. The functionality of the FLINT and GCBM examples are provided through REST APIs written in Flask. 

Reviewer: ```suggestion
A wrapper REST API is created to provide the different endpoints that point to the different operations of the CLI tool. The functionality of the FLINT and GCBM examples are provided through REST APIs written in Flask. 
```

 
## Significance of FLINT Cloud in the ecosystem

> The development of the FLINT Cloud project has been instrumental in leading to the development of projects like [FLINT-UI](https://github.com/moja-global/FLINT-UI) and Data-driven integration with [CML and DVC](https://cml.dev/). The goal of DVC is to bring agility, reproducibility, and collaboration into existing data science workflows. 

Reviewer: ```suggestion
The development of the FLINT Cloud project has been instrumental in the development of projects like [FLINT-UI](https://github.com/moja-global/FLINT-UI) and data-driven integration with [Data version control (DVC) and software Continuous Machine Learning (CML)](https://cml.dev/). The goal of DVC is to bring agility, reproducibility, and collaboration into existing data science workflows. 
```

> FLINT-UI is a frontend interface written in Vue to interact with the FLINT Cloud APIs. This allows people from various technical and non-technical backgrounds to interact with the web application, run simulations and visualize results

Reviewer: ```suggestion
FLINT-UI is a frontend interface written in Vue to interact with the FLINT Cloud APIs. This allows people from various technical and non-technical backgrounds to interact with the web application, run simulations and visualize results.
```

The second project is based on the concepts of reproducible science. 

> Data version control (DVC) and software Continuous Machine Learning (CML) provide a framework to continuously reproduce analyses using Github Actions as well as saving the output of the configuration phase to reuse for subsequent builds. The basic idea is to track the inputs provided to the model and monitor the outputs generated in various simulations over time. Though this feature isn’t natively required for the FLINT cloud project, it provides users with a way to incorporate them in their implementations. 

Reviewer: ```suggestion
DVC and software CML provide a framework to continuously reproduce analyses using GitHub Actions as well as saving the output of the configuration phase for reuse in subsequent builds. The basic idea is to track the inputs provided to the model and monitor the outputs generated in various simulations over time. Though this feature isn’t natively required for the FLINT cloud project, it provides users with a way to incorporate them in their implementations. 
```
